Regex-Support

kimpenhaus opened the pull request:
Hi @realvizu 

I tried to integrate Regex support as requested in #53 - this is by far not perfect and needs at least some documentation on the RegexDomain class itself. I tried to adhere some standards as I forced the regex to have a / delimiter at start and end (similiar to js, perletc)

But the two main problems I have - and which I am asking for some help - are the unit-tests and the nuget/dotnet pack execution.

I am running on an osx machine which seems to struggle with some of the commands:

<img width="4936" alt="grafik" src="https://github.com/user-attachments/assets/df54b93b-48c3-4d2b-83ad-6e65662b3794" />

might be some issues with slashes/backslashes in specific os environments...

Maybe you can give me some support in points of issues - and give me some feedback on what I should change in the codebase.

Thanks.

realvizu commented:
Hi!

First of all, thanks for contributing to NsDepCop!

Regarding the unit tests failing on OSX: that's my bad, I did not pay attention to make the path handling in the tests cross-platform. I've added a commit to your branch, please check out if it solves the problem.

Regarding the nuget/dotnet pack execution: please provide some details what kind of error do you get.

Regarding the feature implementation: I think it's a good idea to use the slash begin/end delimiters for regex rules, so they are easily distingushed from the normal and the wildcard rules. My main concern with regex rules is the possible performance impact, and this implementation ensures that the normal and wildcard rules are not affected in any way, so it's great.

Regarding the GetMatchRelevance method implementation: its purpose is to find the "best" matching rule in case there are multiple matching ones. Currently the Domain class (that encapsulates a concrete namespace) is considered highest precedence, followed by the WildcardDomain, which returns a lower GetMatchRelevance value, based on how many wildcard substitutions were needed to make a match. In order to be backward compatible, the RegexDomain should be the last in the rules precedence order, so it should return a GetMatchRelevance value which is always smaller then the ones given by Domain and WildcardDomain. I'm not sure how well I explained it so feel free to ask for more clarification is needed. Also, please add some unit tests that check that in case both a WildcardDomain and a RegexDomain rule gives a match, then the WildcardDomain is chosen as best matching.

kimpenhaus commented:
Hi @realvizu 

so I found some time to updarte the PR - I think I addressed most of the poiunts you mentioned - heopfully 😄 

1) performance - I've also added an timeout to the regex parsing which should omit excessively long execution

2) relevance - I think I understood what you've meant and moved the regex match relevance to the bottom last ... at least I've tried. as I think there is only match or no match I used 1 and 0

3) unit tests - I've added some basic tests

What I am still missing and would need some point of direction:

> Also, please add some unit tests that check that in case both a WildcardDomain and a RegexDomain rule gives a match, then the WildcardDomain is chosen as best matching.

Could you give me some hints where to put those tests and what they should look alike? some basic structure I could follow is more than sufficient.

thanks 🙏🏼

realvizu commented:
Hi,

Thanks again for your work, I've added a commit to your branch with the following changes.


1. Unit tests

Added TypeDependencyValidatorTests.AllowRule_WildcardRuleIsStrongerThanRegexRule test case to validate that if both a WildcardRule and a RegexRule matches, then the WildcardRule is used.
This is important only in the edge case, when the matched rule also contains an allowed type name list, but in this case it is important to prioritize WildcardRule to be backward compatible.

Refactored DomainSpecificationTests to be more clear that we test two different things:
- that GetMatchRelevance returns the expected value (for many different kinds of rules),
- and that GetMatchRelevance implements this priority order: SimpleRule, WildCardRule, RegexRule.

Added RegexDomainTests.GetMatchRelevance_WhenTimeoutOccurs_ShouldReturnNoMatch_WithoutThrowingException.


2. Catching RegexMatchTimeoutException

You've added a Regex timeout value, which is fine to avoid excessively long processing, but we also have to catch the RegexMatchTimeoutException that it can throw.
So in the case when the Regex matching takes too much time, we treat it as a non-matching rule.


3. Performance

I was curious how the performance of the Regex rules compare to the simple and wildcard rules, so I've added a benchmark.
See the NsDepCop.Benchmarks project, and the [result file](https://github.com/kimpenhaus/NsDepCop/blob/master/source/NsDepCop.Benchmarks/NsDepCop.Benchmarks.RuleTypesBenchmarks-report-github.md).

It turns out that it's a tricky thing, and depends very much on the usage pattern.

When using the static Regex.IsMatch method, the engine caches the last 15 most recently used static regular expression patterns (see: https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-regex#use-time-out-values), so it's not suitable for our usage pattern, because we can easily have more than 15 regex patterns in an NsDepCop config, so caching won't be effective, and performance will drop. 
To test the performance impact of caching, I also run the benchmarks with the Regex cache turned off. It's really subtantial.

So we have to create our own Regex object instance in the RegexDomain class, and use that, effectively caching our regex patterns by ourselves.

There's also the RegexOptions.Compiled flag, which incurs a startup cost, when the regex pattern is compiled to IL, and then it's faster than the interpreted mode.
So it depends on the actual regex pattern, and on how many times you use it, that it's worth compiling the regex or not.
In our benchmarks, it turns out that it's not worth compiling.

To make these benchmarks possible, I've added some flags to the RegexDomain ctor that controls the behavior, and set the defaults to the combination that performed best in the benchmarks: 
- RegexUsageMode.Instance 
- RegexCompilationMode.Interpreted


Please let me know what you think about these changes and the perf analysis.
Do you see any loose ends that we should tie up before releasing the new version?

kimpenhaus commented:
oh wow - thanks for your effort. 👍🏼  that wasn't my intention when I asked for some guidance but highly appreciate it.

good catch with the timeout - my fault I haven' thought about the execption. 

I think its worth giving it a release and see how it behaves.

thanks again for your time and help 🙏🏼

realvizu commented:
Merged, soon to be release.

realvizu commented:
Released: https://www.nuget.org/packages/NsDepCop/2.6.0

Thanks for your contribution!

kimpenhaus commented:
From my perspective you did the most of work ☺️ thanks!
